provers/README.md

```diff
@@ -5,3 +5,98 @@ Directory to place source code of prover clients of verifier contracts.
+$ pip3 install tensorflowjs
+
+# convert model from keras format to tf format (e.g. converting demo model)
+$ yarn convert-keras-to-tf demo demo-tfjs
+```
+
+Also, make sure that hardhat is running on localhost, and three contracts (`Verifier`, `CustomVerifier`, `CustomVerifierFactory`) have already been deployed. 
+
+```sh
+# cd into 'contracts' directory
+$ cd ../contracts
+
+# start hardhat blockchain in one window (FIRST TIME ONLY)
+$ yarn run-hardhat
+
+# open another window & deploy contracts (FIRST TIME ONLY)
+$ yarn deploy-contracts localhost
+```
+
+## 1. Encoding Testing Results
+
+We first would like to make predictions of the model against private testing data. Run the following commands to encode testing results as a Merkle tree, where each argument refers to the following.
+
+| Argument Name | Description | Default |
+|--------------:|-------------|---------|
+| **MODEL** | name of model directory | `demo-tfjs` |
+| **MERKLE_TREE** | name of `.json` file containing Merkle tree | `demo-tree` |
+
+```sh
+$ yarn encode-testing-results --MODEL=${aaa} --MERKLE_TREE=${bbb}
+```
+
+## 2. Commit Merkle Root
+
+We are making use of commit-reveal scheme to avoid modifications of the previous commitments. Run the following commands to commit the Merkle root to our custom verifier contract.
+
+```sh
+# cd into 'contracts' directory
+$ cd ../contracts
+
+# commit Merkle root to the custom verifier
+$ yarn commit-merkle-root localhost
+
+# confirm commitment
+$ yarn get-commitments localhost
+```
+
+## 3. Generating Merkle Proofs
+
+Run the following commands to generate Merkle proofs of leaf nodes matched with the generated random challenge, where each argument refers to the following.
+
+| Argument Name | Description | Default |
+|--------------:|-------------|---------|
+| **MERKLE_TREE** | name of `.json` file containing Merkle tree | `demo-tree` |
+| **CIRCUIT_INPUT** | name of directory containing circuit inputs | `demo` |
+
+```sh
+$ yarn generate-merkle-proofs --MERKLE_TREE=${aaa} --CIRCUIT_INPUT=${bbb}
+```
+
+## 4. Generating ZKPs
+
+Run the following commands to generate ZKPs as well as actual inputs to be verified by the verifier contract auto-generated by `Circom`, where each argument refers to the following.
+
+| Argument Name | Description | Default |
+|--------------:|-------------|---------|
+| **CIRCUIT_EXECUTABLE** | name of circuit's executable files | `demo-circuit` |
+| **ZKEY_FINAL** | name of final `.zkey` file | `demo_0001` |
+
+```sh
+# generate ZKPs
+$ yarn generate-zkps --CIRCUIT_EXECUTABLE=${aaa} --ZKEY_FINAL=${bbb}
+
+# generate parameters for verifier contract
+$ yarn parameterize-zkps
+```
+
+## 5. Send Merkle Proofs & ZKPs to Verifier for Verifications
+
+Run the following commands to send Merkle proofs and ZKPs to our custom verifier contract for verification. **Note that you need to modify solidity version of the auto-generated verifier to `0.8.19` before running those commands.** 
```

Comment:
This is a bit annoying, but shouldn't be that problematic (if we lower solidity version, we also have to lower versions of dependencies already pinned).